### PR TITLE
Added support for batched prefill warmup in TTT + added creation of sampling params for warmup prefill for Llama70b

### PR DESCRIPTION
**Pull request by nostojicTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/37732

### Problem description
- Added support for batched prefill warmup for TTT. When batched prefill gets merged to main, just uncomment the `batch==32` if statement.
- Changed Llama70b prefill warmup such that all possible prefill sampling parameters are created.
- Changed that all combinations of sampling parameters are sweeped once in prefill. That means that instead of running all combinations of sampling parameters for each sequence length, we will sweep them just once withing the first sequence length, since the sampling module is sequence length agnostic

- When adding the sweep of sampling parameters in prefill, these are the new time needed for prefill warmup to complete (Llama70b on galaxy with dp=4):

| Before  | After  |
|---------|--------|
| 2m 33s  | 3m 13s |

vllm nightly - https://github.com/tenstorrent/tt-metal/actions/runs/22346700092
t3k demo tests - https://github.com/tenstorrent/tt-metal/actions/runs/22346708909
single card demo tests - https://github.com/tenstorrent/tt-metal/actions/runs/22346718753
galaxy demo tests - https://github.com/tenstorrent/tt-metal/actions/runs/22346728503
blackhole demo tests - https://github.com/tenstorrent/tt-metal/actions/runs/22346740144

### CODE REVIEW

File: models/demos/llama3_70b_galaxy/tt/llama_model.py [hunk 1]
```diff
@@ -187,8 +187,10 @@ def setup_decode(self, mesh_sub_device_manager_id_decode=None):
                 mesh_device=self.mesh_device,
                 tt_ccl=self.tt_ccl,
             )
+            self._supports_on_device_sampling = True
         else:
             self.tt_ccl = self.tt_ccl_decode
+            self._supports_on_device_sampling = False
```
Comment (mtairum) [inline models/demos/llama3_70b_galaxy/tt/llama_model.py]: Is this supposed to be false here?

This if/else is checking if the decode already exists. If not it does setup of decode CCLs and sampling. If it exists will just points the `self.tt_ccl` to the decode one. why do we want to make device sampling false then?

**Review by mtairum — CHANGES_REQUESTED** [review]

File: tt_metal/third_party/tt_llk
Comment (fvranicTT) [inline tt_metal/third_party/tt_llk:1]: Please drop this change. It seems unrelated.

Comment (nostojicTT) [reply tt_metal/third_party/tt_llk:1]: accidentally pushed

File: models/tt_transformers/tt/generator.py [hunk 2]
```diff
@@ -161,42 +159,65 @@ def warmup_model_prefill(self, kv_cache, enable_trace, can_sample_on_device, non
                 ):
                     continue
 
-                warmup_tokens = torch.zeros(1, supported_length, dtype=torch.long)
-                warmup_prompt_lens = torch.tensor([supported_length], dtype=torch.long)
-                warmup_empty_slots = list(range(1))
+                for batch_size in [1, 32]:
+                    if batch_size == 32:
+                        logger.warning("Batched prefill in TTT is not supported")
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: Should just log once instead of in each outer loop iter to avoid spamming warnings

File: models/tt_transformers/tt/generator.py [hunk 3]
```diff
@@ -161,42 +159,65 @@ def warmup_model_prefill(self, kv_cache, enable_trace, can_sample_on_device, non
...
+                            can_sample_on_device=can_sample_on_device,
+                            non_greedy_decoding_on_device=non_greedy_decoding_on_device,
+                            batch_size=batch_size,
+                        )
+                    else:
+                        sampling_params = [None]
+
+                    for param in sampling_params:
+                        logger.info(
+                            f"Warming up prefill for sequence length: {supported_length} with sampling params: {param}"
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py]: maybe add batch size to this log info and remove the other logger info above

File: models/tt_transformers/tt/generator.py [hunk 4]
```diff
@@ -161,42 +159,65 @@ def warmup_model_prefill(self, kv_cache, enable_trace, can_sample_on_device, non
...
+                            page_table_warmup,
+                            kv_cache,
+                            warmup_prompt_lens,
+                            warmup_empty_slots,
+                            enable_trace,
+                            model_id,
+                            param,
+                        )
+
+                    sampling_parameters_sweeped = True
```
Comment (skhorasganiTT) [inline models/tt_transformers/tt/generator.py:162]: this should be outside the batch size loop right? (i.e. left indented once?)

**Review by skhorasganiTT — APPROVED** [review]

File: models/tt_transformers/tt/generator.py (hunk 4 above)
Comment (nostojicTT) [reply models/tt_transformers/tt/generator.py:162]: It should be here because right after we sweeped all sampling parameters for batch-1 there is no reason to sweep them for batched prefill also.

Comment (skhorasganiTT) [reply models/tt_transformers/tt/generator.py:162]: Wouldn't the sampling operation be different for different batch sizes? cc @sraizada-tt

Comment (sraizada-tt) [reply models/tt_transformers/tt/generator.py:162]: no, prefill loops over batch-1 for sampling

File: models/demos/llama3_70b_galaxy/tt/generator.py [hunk 5]
```diff
@@ -27,6 +27,15 @@
 from models.common.warmup import WarmupForwardMixin
 
 
+def get_prefill_warmup_sequence_lengths(max_seq_len: int) -> list[int]:
+    """
+    Returns powers of 2 from 128 up to max_seq_len (inclusive).
+    """
+    return ([128] if max_seq_len >= 128 else []) + [
```
Comment (mtairum) [inline models/demos/llama3_70b_galaxy/tt/generator.py]: shouldn't 128 be returned always? 

If for some reason your prompt is <128 and you define your max_seq_len < 128, you'll return an empty list. 
Also max_seq_len should always be >= 128, so this is a bit redundant.

**Review by mtairum — APPROVED** [review]

File: models/demos/llama3_70b_galaxy/tt/generator.py (hunk 5 above)
Comment (nostojicTT) [reply models/demos/llama3_70b_galaxy/tt/generator.py]: I agree that it is redudant but I would like to keep this here if somehow `self.args.max_seq_len` is set incorrectly. (max_seq_len argument will always be `self.args.max_seq_len`)

Comment (mtairum) [reply models/demos/llama3_70b_galaxy/tt/generator.py]: But we have other checks for max_seq_len in other parts of the code. even if max_seq_len is set to below 128 for some reason and that passes the code up to this point, we still want to warmup up to 128, since that's the minimum prefill length we compute.

Comment (nostojicTT) [reply models/demos/llama3_70b_galaxy/tt/generator.py]: Alright, will set it to cover 128 always
